src/main/kotlin/com/sourcegraph/cody/context/RemoteRepoInsight.kt

```diff
@@ -0,0 +1,394 @@
+          else -> Icons.RepoHostGeneric
+        }
+}
+
+val RemoteRepoLanguage = object : Language("SourcegraphRemoteRepoList") {}
+
+class RemoteRepoFileType : LanguageFileType(RemoteRepoLanguage) {
+  companion object {
+    @JvmStatic val INSTANCE = RemoteRepoFileType()
+  }
+
+  override fun getName(): String {
+    return "SourcegraphRemoteRepoListFile"
+  }
+
+  override fun getDescription(): String {
+    return "A list of Sourcegraph repository indexes"
+  }
+
+  override fun getDefaultExtension(): String {
+    return ""
+  }
+
+  override fun getIcon(): Icon? {
+    return null
+  }
+}
+
+class RemoteRepoTokenType(debugName: @NonNls String) : IElementType(debugName, RemoteRepoLanguage) {
+  override fun toString(): String {
+    return "RemoteRepoTokenType." + super.toString()
+  }
+
+  companion object {
+    val REPO = RemoteRepoTokenType("REPO")
+    val SEPARATOR = RemoteRepoTokenType("SEPARATOR")
+    val EOF = RemoteRepoTokenType("EOF")
+  }
+}
+
+class RemoteRepoFile(viewProvider: FileViewProvider?) :
+    PsiFileBase(viewProvider!!, RemoteRepoLanguage) {
+  override fun getFileType(): FileType {
+    return RemoteRepoFileType.INSTANCE
+  }
+
+  override fun toString(): String {
+    return "Sourcegraph Remote Repo File"
+  }
+}
+
+private enum class LexerState(val value: Int) {
+  IN_REPO(1),
+  IN_SEPARATOR(2),
+  EOF(3);
+
+  companion object {
+    fun fromInt(value: Int): LexerState? = values().find { it.value == value }
+  }
+}
+
+internal class RemoteRepoListParserDefinition : ParserDefinition {
+  override fun createLexer(project: Project): Lexer {
+    return object : Lexer() {
+      var buffer: CharSequence = ""
+      var startOffset: Int = 0
+      var endOffset: Int = 0
+      var state: LexerState = LexerState.EOF
+      var offset: Int = 0
+
+      override fun start(
+          buffer: CharSequence,
+          startOffset: Int,
+          endOffset: Int,
+          initialState: Int
+      ) {
+        this.buffer = buffer
+        this.startOffset = startOffset
+        this.endOffset = endOffset
+        offset = startOffset
+        state = LexerState.fromInt(initialState) ?: stateAtOffset()
+      }
+
+      override fun getState(): Int {
+        return this.state.value
+      }
+
+      override fun getTokenType(): IElementType? {
+        return when (state) {
+          LexerState.IN_REPO -> RemoteRepoTokenType.REPO
+          LexerState.IN_SEPARATOR -> RemoteRepoTokenType.SEPARATOR
+          LexerState.EOF -> null
+        }
+      }
+
+      override fun getTokenStart(): Int {
+        return this.offset
+      }
+
+      override fun getTokenEnd(): Int {
+        val index =
+            when (tokenType) {
+              RemoteRepoTokenType.REPO ->
+                  buffer.indexOfAny(charArrayOf(' ', '\t', '\r', '\n'), offset)
+              RemoteRepoTokenType.SEPARATOR -> {
+                val subIndex =
+                    buffer.subSequence(offset, buffer.length).indexOfFirst { ch ->
```

Comment:
In case of `\r\n` do we care about index of `\r`, `\n`, or doesn't really matter?
Because in case of `\r\n` it does not really return index of the last char in line.